Content.Server/Abilities/Psionics/PsionicAbilitiesSystem.cs

```diff
@@ -62,66 +77,185 @@ public void AddPsionics(EntityUid uid, bool warn = true)
                 AddRandomPsionicPower(uid);
         }
 
-        public void AddPsionics(EntityUid uid, string powerComp)
+        public void AddRandomPsionicPower(EntityUid uid)
         {
-            if (Deleted(uid)
-                || HasComp<PsionicComponent>(uid))
+            EnsureComp<PsionicComponent>(uid, out var psionic);
+
+            if (!_prototypeManager.TryIndex<WeightedRandomPrototype>("RandomPsionicPowerPool", out var pool))
+            {
+                _sawmill.Error("Can't index the random psionic power pool!");
                 return;
+            }
+
+            var newPool = _serialization.CreateCopy(pool, null, false, true);
+            foreach (var proto in pool.Weights.Keys)
+            {
+                if (!_prototypeManager.TryIndex<PsionicPowerPrototype>(proto, out var powerPrototype))
+                    continue;
 
-            AddComp<PsionicComponent>(uid);
+                if (psionic.ActivePowers.Contains(powerPrototype))
+                    newPool.Weights.Remove(powerPrototype.ID);
+            }
 
-            var newComponent = (Component) _componentFactory.GetComponent(powerComp);
-            newComponent.Owner = uid;
+            if (newPool.Weights.Keys != null)
+            {
+                var newPower = _prototypeManager.Index<PsionicPowerPrototype>(newPool.Pick());
+                InitializePsionicPower(uid, newPower);
+            }
 
-            EntityManager.AddComponent(uid, newComponent);
+            _glimmerSystem.Glimmer += _random.Next(1, 5);
```

Comment:
It's probably a good idea to turn these values into consts at the top of the file, to make it easier to modify these values, if that is ever desired in the future.